src/2017/subbehaviors.org

```diff
@@ -0,0 +1,150 @@
+    Defense::defending
+        SubmissiveDefender::marking
+            Move::completed[robot=5]
+                face(0.160718, 9.15305)
+        SubmissiveDefender::marking
+            Move::completed[robot=3]
+                face(0.160718, 9.15305)
+        SubmissiveGoalie::block
+            Move::running[robot=0]
+                face(0.160718, 9.15305)
+                move(0.290916, 0.14)
+                endVelocity(0, 0)
+#+END_SRC
+
+*** Details                                                          :docs:
+- Pretty much same thing here
+- We've got a Defense tactic running, with SubmissiveDefender and Goalie Skills
+- These skills actually run robot commands.
+* How do I use subbehaviors?
+** Adding Subbehaviors
+#+BEGIN_SRC python
+  self.add_subbehavior(skill_object,
+                       "Name of Subbehavior",
+                       required=False or True
+                       priority=10) # A higher number is higher priority
+#+END_SRC
+
+*** Details                                                          :docs:
+- For more docs on this entire section see [[https://robojackets.github.io/robocup-software/classgameplay_1_1single__robot__composite__behavior_1_1_single_robot_composite_behavior.html][this link.]]
+- These subbehaviors show up in the behavior tree when you run your program.
+- This can be extremely useful when debugging state transitions or subbehavior assignments.
+** Removing Subbehaviors
+#+BEGIN_SRC python
+self.remove_subbehavior('string name')
+
+self.remove_all_subbehaviors()
+#+END_SRC
+** Getting Subbehavior Plays
+#+BEGIN_SRC python
+a_subbheavior = self.subbehavior_with_name('string name')
+#+END_SRC
+* Real Examples
+** CoordinatedPass Tactic
+#+BEGIN_SRC python
+  def on_enter_running(self):
+      receiver = skills.pass_receive.PassReceive()
+      receiver.receive_point = self.receive_point
+      self.add_subbehavior(receiver,
+                           'receiver',
+                           required=self.receiver_required)
+
+  def on_exit_running(self):
+      self.remove_subbehavior('receiver')
+#+END_SRC
+** Line Up Tactic
+- First State Machine is set up, then:
+#+BEGIN_SRC python
+  # Triggered whenever the line changes
+  self.remove_all_subbehaviors()
+  for i in range(6):
+      pt = self._line.get_pt(0) + (self.diff * float(i))
+      self.add_subbehavior(
+          skills.move.Move(pt),
+          name="robot" + str(i),
+          required=False,
+          priority=6 - i)
+#+END_SRC
+
+#+BEGIN_SRC python
+  def execute_running(self):
+      for i in range(6):
+          pt = self._line.get_pt(0) + (self.diff * float(i))
+          self.subbehavior_with_name("robot" + str(i)).pos = pt
+#+END_SRC
+* RoboCup Pro Tip
+- Find some code doing something like what you want
+- Tweak it until it works
+- It's less effective than working everything out, but it's great for beginners!
+
+* Assignment
+- Create a Triangle Pass Play
+- Round up 3 Robots in a triangle, and pass between them.
```

Comment:
It might be a little confusing to use `Round up` here instead of something a little more bland like `Move`